tests/ga/test_update.py

```diff
@@ -1740,6 +1740,34 @@ def test_it_should_recreate_extension_event_directories_for_existing_extensions_
                 for ext_dir in expected_events_dirs:
                     self.assertTrue(os.path.exists(ext_dir), "Extension directory {0} should exist!".format(ext_dir))
 
+    def test_it_should_wait_to_fetch_first_goal_state(self):
+        with _get_update_handler() as (update_handler, protocol):
+            with patch("azurelinuxagent.common.logger.warn") as patch_warn:
+                with patch("azurelinuxagent.common.logger.info") as patch_info:
+
+                    # Fail GS fetching for the 1st 5 times the agent asks for it
+                    update_handler._fail_gs_count = 5
+
+                    def get_handler(url, **kwargs):
+                        if HttpRequestPredicates.is_goal_state_request(url) and update_handler._fail_gs_count > 0:
+                            update_handler._fail_gs_count -= 1
+                            return MockHttpResponse(status=500)
+                        return protocol.mock_wire_data.mock_http_get(url, **kwargs)
+
+                    protocol.set_http_handlers(http_get_handler=get_handler)
+                    update_handler.run(debug=True)
+
+        self.assertTrue(update_handler.exit_mock.called, "The process should have exited")
+        exit_args, _ = update_handler.exit_mock.call_args
+        self.assertEqual(exit_args[0], 0, "Exit code should be 0; List of all warnings logged by the agent: {0}".format(
+            patch_warn.call_args_list))
+        warn_msgs = [args[0] for (args, _) in patch_warn.call_args_list if
+                     "An error occurred while retrieving the goal state" in args[0]]
+        self.assertTrue(len(warn_msgs) > 0, "Error should've been reported when failed to retrieve GS")
```

Comment:
Since we are failing it exactly 5 times , will it be beneficial to check the`len(warn_msgs)` to be equal to 5?

Reply:
that would be ideal if we were actually logging a warn each time. The `_try_update_goal_state` function only logs it once (might be worth checking out the implementation of it). 

Anywho, removed that logic